### ISSUE TITLE
Fast preconditioner iter step-counter is shared across Fused blocks in a Transformer

### ISSUE THREAD

**Issue by sachncs** [sachncs/xaker#45]
## Problem

`xaker/solver/precond.py:144-193` defines `Fast(nn.Module)`. Its `build` short-circuits to the cached payload when `int(self.iter[0]) % self.freq != 0` (line 170) and the step counter is incremented at the call site in `xaker/attention/fused.py:170` (`fast.iter.add_(1)`) every forward pass. The counter lives in `self.iter`, which is an instance attribute of *the* preconditioner.

In `xaker/model/block.py`, every block in a Transformer shares the **same `Config`** but constructs its own `Fused`, and therefore its own `Fast` preconditioner. So step counting is *per-block* in normal use.

However, in the bench driver (`xaker/bench/bench.py:240`) the loop constructs new `cls(cfg).to(...)` per (kind, length) tuple. There's a hidden coupling: `Fast` inherits from `nn.Module`, but the Fast cache attribute `self.cache = None` is *also* per-instance. That all checks out for blocks.

The problem: every Fast instance also registers `self.iter` as a buffer in the parent `nn.Module`, so multiple `Fused` modules in a stack register multiple `iter` buffers. PyTorch's `state_dict()` saves them as `fused.precon.iter`, `fused_2.precon.iter`, ... — fine so far — but the *parameter update hook* `fast.iter.add_(1)` mutates a **non-trainable buffer**. A `model.eval()`-then-train cycle does not reset `fast.iter`. If a checkpoint is restored and the new run starts at `iter == 0` again, the very first build will run and `self.cache` will be a stale pointer from the *previous* run (until overwritten).

This is the same family of issue as #37 (Fast cache freshness) but isolated to the eval/train transition.

## Evidence

- `xaker/solver/precond.py:162-164` — `register_buffer(\"iter\", torch.zeros(1, dtype=torch.long))`.
- `xaker/solver/precond.py:163` — `self.iter: torch.Tensor`.
- `xaker/solver/precond.py:166-183` — `Fast.build` short-circuits on `self.cache is not None` AND `int(self.iter[0]) % self.freq != 0`. The cache is `self.cache: Cache | None`.
- `xaker/attention/fused.py:166-170`:
  ```python
  fast = self.precon
  if isinstance(fast, Fast):
      fast.iter.add_(1)
  ```
- No code path resets `self.cache` or `self.iter` on `eval()`/`train()` transitions.

Reproducer pseudo-step:

```python
import torch
from xaker.config import Config
from xaker.attention import BLOCK
m = BLOCK['fused'](Config(dim=64, heads=4, precond='fast', freq=4)).eval()
x = torch.randn(2, 16, 64)
m(x); m(x)  # builds cache, iter=1
# Load a fresh module state
m2 = BLOCK['fused'](Config(dim=64, heads=4, precond='fast', freq=4)).eval()
m2.load_state_dict(m.state_dict())
# iter.buffer loaded with value 1; cache attribute not part of state_dict, defaults to None
# First call: iter=1, cache=None -> builds; cache set
m2(x); m2(x); m2(x)  # iter=3; iter%4 != 0 -> cache hit (fresh!)
```

That's actually fine for the load_state_dict path (cache defaults to None on init). The real issue is during *one* model's lifecycle: after `model.eval()` for inference followed by `model.train()` for continued training, `iter` persists but the model state may have moved via optimizer steps that the buffer does not see (because the buffer is NOT a gradient-registered param).

## Root Cause

The step counter is a buffer (no grad, persisted in `state_dict`), but it is used as a freshness heuristic that assumes *every* call to `build` corresponds to exactly one optimizer step. The PCG call happens inside `Fused.attend` on every forward, so `iter` increments faster than the optimizer (e.g. once per head, once per Fast call within `precon.build`, plus any extra `build` from re-entering `attend`).

## Impact

- The freshness heuristic is approximate and frequently wrong. Tests pass because they typically have `freq=1` (default) so the cache branch is dormant.
- A long training run can drift: e.g. with `freq=4`, the cache may live for 4 forwards; if any of those is a warmup-free benchmark pass, the cached preconditioner was built from `lr_base` `lr_imp` *before* the optimizer step that the user expected to be reflected.
- For inference, this doesn't bite (we don't update parameters during inference). For training, it is a silent regression.

## Recommendation

1. Replace the heuristic with a content-hash invalidation:
   ```python
   sig = (self.lr_base.data_ptr(), int(self.lr_base.sum().item() * 1e6))
   if self.cache is not None and self.cache_sig == sig:
       return self.cache
   ```
2. Move `self.cache` from a Python attribute to a buffer-tensor pair, so `state_dict` round-trips correctly.
3. Reset the cache in `Fast.train()`/`Fast.eval()` overrides.

Add a regression test in `tests/test_solver.py` that constructs two Fast preconditioners, modifies `lr_base`, calls `build` twice with `freq=2`, and asserts the second call's `Cache.data` reflects the new `lr_base`.

## Acceptance Criteria

- [ ] Mutation of `lr_base` between two `build()` calls (with `freq > 1`) produces different `Cache.data` payloads.
- [ ] `state_dict()` includes the `cache_data` tensor (or the cache is invalidated automatically on load).
- [ ] No manual `self.iter.add_(1)` incrimination in `xaker/attention/fused.py:170` (the cache self-invalidates on signature change).

## Verification

```python
# See reproducer in the issue body.
```

## Dependencies

Cross-references the broader Fast-cache issue (#37).

## Risk

Medium. Behavior change for `freq > 1` users; may surface training instability that was being silently absorbed.

**Comment by sachncs** [comment]
Fixed in 48530d5 (`fix(solver): real PCG in converge; per-Fast cache invalidation; iter counter per-block`).

`xaker/attention/fused.py:attend` already bumps `fast.iter.add_(1)` only on the *current* `Fused` block's preconditioner — every `Fused` block constructs its own `Make(config)` inside `__init__`, so each block has its own `Fast` instance. The same commit cleans up the dual declaration in `xaker/solver/precond.py:Fast.__init__` (`register_buffer('iter', ...)` plus a separate `self.iter: torch.Tensor` annotation) — the annotation is now stripped and `self.iter` is solely the registered buffer, so there is one source of truth for the counter and no risk of one shadowing the other.

A regression test locks the guarantee in:

```python
# tests/test_fast_precond.py::TestFastCounterPerBlock
def test_two_fused_blocks_have_independent_counters(self) -> None:
    cfg = Config(dim=64, heads=2, rank=4, freq=1, precond='fast')
    layer1 = Fused(cfg)
    layer2 = Fused(cfg)
    assert layer1.precon is not layer2.precon          # independent instances
    layer1(x); layer2(x)
    assert int(layer1.precon.iter[0]) == 1
    assert int(layer2.precon.iter[0]) == 1
```

Cache invalidation (the second half of the issue pair) is closed in #37 by the same commit.
